archiver/worker.go

```diff
@@ -0,0 +1,264 @@
+		if _, err := gzWriter.Write(path, append(j, '\n')); err != nil {
+			return "", fmt.Errorf("failed to write to file: %w", err)
+		}
+	}
+	if err := gzWriter.Close(); err != nil {
+		return "", fmt.Errorf("failed to close file: %w", err)
+	}
+	defer func() { _ = os.Remove(path) }()
+
+	fileUploader, err := w.storageProvider.GetFileManager(workspaceID)
+	if err != nil {
+		log.Errorw("Skipping storing errors since no file manager is found", "error", err)
+		return "", fmt.Errorf("failed to get file manager: %w", err)
+	}
+
+	file, err := os.Open(path)
+	if err != nil {
+		return "", fmt.Errorf("failed to open file %s: %w", path, err)
+	}
+	defer func() { _ = file.Close() }()
+	prefixes := []string{
+		w.sourceID,
+		w.archiveFrom,
+		firstJobCreatedAt.Format("2006-01-02"),
+		fmt.Sprintf("%d", firstJobCreatedAt.Hour()),
+		w.config.instanceID,
+	}
+	uploadOutput, err := fileUploader.Upload(ctx, file, prefixes...)
+	if err != nil {
+		log.Errorw("failed to upload file to object storage", "error", err)
+		return "", fmt.Errorf("failed to upload file to object storage - %w", err)
+	}
+
+	return uploadOutput.Location, nil
+}
+
+func (w *worker) getJobs() ([]*jobsdb.JobT, bool, error) {
+	defer w.fetchLimiter.Begin(w.sourceID)()
+	params := w.queryParams
+	params.PayloadSizeLimit = w.payloadLimitFunc(w.config.payloadLimit())
+	params.EventsLimit = w.config.eventsLimit()
+	params.JobsLimit = w.config.eventsLimit()
+	unProcessed, err := w.jobsDB.GetUnprocessed(w.lifecycle.ctx, params)
+	if err != nil {
+		w.log.Errorw("failed to fetch unprocessed jobs for backup", "error", err)
+		return nil, false, err
+	}
+	return unProcessed.Jobs, unProcessed.LimitsReached, nil
+}
+
+func marshalJob(job *jobsdb.JobT) ([]byte, error) {
+	var J struct {
+		UserID       string          `json:"userId"`
+		EventPayload json.RawMessage `json:"payload"`
+		CreatedAt    time.Time       `json:"createdAt"`
+		MessageID    string          `json:"messageId"`
+	}
+	J.UserID = job.UserID
+	J.EventPayload = job.EventPayload
+	J.CreatedAt = job.CreatedAt
+	J.MessageID = gjson.GetBytes(job.EventPayload, "messageId").String()
+	return json.Marshal(J)
+}
+
+func (w *worker) markStatus(
+	jobs []*jobsdb.JobT, state string, response []byte,
+) error {
+	defer w.updateLimiter.Begin(w.sourceID)()
+	return misc.RetryWithNotify(
+		w.lifecycle.ctx,
+		w.config.uploadFrequency,
+		w.config.jobsdbMaxRetries(),
+		func(ctx context.Context) error {
+			return w.jobsDB.UpdateJobStatus(
+				ctx,
+				lo.Map(jobs, func(job *jobsdb.JobT, _ int) *jobsdb.JobStatusT {
+					return &jobsdb.JobStatusT{
+						JobID:         job.JobID,
+						JobState:      state,
+						ErrorResponse: response,
+						AttemptNum:    job.LastJobStatus.AttemptNum + 1,
+						ExecTime:      time.Now(),
+						RetryTime:     time.Now(),
+					}
+				}),
+				nil,
+				nil,
+			)
+		},
+		func(attempt int) {
+			w.log.Warnw("failed to mark jobs' status", "attempt", attempt)
+		},
+	)
+}
+
+func errJson(err error) []byte {
```

Comment:
[nit-pick] idiomatic way in go
```suggestion
func errJSON(err error) []byte {
```